src/natcap/invest/wave_energy.py

```diff
@@ -619,9 +618,9 @@ def execute(args):
         output_dir, 'GridPts_prj%s.shp' % file_suffix)
 
     grid_data = grid_land_data.loc[
-        grid_land_data['TYPE'].str.lower() == 'grid']
+        grid_land_data['type'].str.lower() == 'grid']
     land_data = grid_land_data.loc[
-        grid_land_data['TYPE'].str.lower() == 'land']
+        grid_land_data['type'].str.lower() == 'land']
```

Comment:
Would the `.str.lower()`s here be subsumed by the `to_lower` parameter?

Reply:
This raises another point, which is should all string data in the CSV be converted to lowercase, or just the column headers? Right now `read_csv_to_dataframe()` is only modifying the column headers, so the `.lower()` here would still be needed, but I think it could make sense to apply it to the entire dataframe

Reply:
That _is_ an interesting question.  Usually we like to use a rule of 3 for creating general solutions, to avoid premature optimization, and to make sure we're developing abstractions that will definitely be used.  In this case, do we have 3 cases in InVEST where we benefit from case-insensitive string comparisons in InVEST?

Reply:
It looks like `utils.build_lookup_from_csv()` does make the entire dataframe lowercase by default, which must be used in several places. And there are definitely at least 3 places in this PR where the case of the column names matter. The case of the values matters here, in wind_energy.py below, and in hra.py: [https://github.com/natcap/invest/pull/259/files#diff-ef8666f31ebc9043012bf1e7a9abadd5R2480](url), and likely several other places.

Reply:
I don't doubt that a third one exists, but I'd also suggest that we confirm it before abstracting the functionality into the proposed helper function.

You do also make a good point that `build_lookup_from_csv` also provides this, and we could offer it in this new function in the interest of consistency.

Reply:
Here's a summary:

Table values should be made lowercase:
- Crop Production (so crop names match those in accepted list)
- Wave energy (grid connection points TYPE value)
- Wind energy (grid connection points TYPE value)

Table values should not be made lowercase:
- HRA Habitat and Criteria CSVs (contains filepaths)
- Forest Carbon Edge Effect (keep case in description field)
- Recreation client (contains filepaths)

Doesn't matter:
- Carbon
- Coastal blue carbon
- Pollination
- SWY
- NDR
- SDR
- Urban cooling
- Urban flood risk
- Habitat quality
- Wave energy (other tables)
- Wind energy (other tables)

Scenic Quality: No CSV input.

Finfish, Fisheries, Recreation server: Reads CSVs in a specialized way line by line. It seems best not to mess with this.

Reply:
My conclusions:
- Making the column names lowercase is always good
- Making all the values lowercase is never necessary, it is only needed for a couple of columns in a couple of inputs.
- Therefore, the `to_lower` arg to `build_lookup_from_csv` should be changed so it only affects the column names, not the entire dataframe. `.lower()`ing of the data should be done in the model only where necessary.
- We could do away with the `to_lower` arg for `build_lookup_from_csv` and `read_csv_to_dataframe` entirely, and have them always make the column names lowercase.

Reply:
> Making the column names lowercase is always good

It's not about a value statement, it's about the problem we're trying to solve.  And in this case, the problem we're trying to solve here is that we want to handle column names in a case-insensitive way and (ideally) consistently across models.  To that end, using lowercase column names sounds like a good convention.

> Making all the values lowercase is never necessary, it is only needed for a couple of columns in a couple of inputs.

You're right, I don't think we can safely lowercase _all_ string table values, and we've probably just gotten lucky with the ones that have been lowercased!  Case will often matter (depending on the OS) for filepaths provided via a table, for example.

> Therefore, the `to_lower` arg to `build_lookup_from_csv` should be changed so it only affects the column names, not the entire dataframe. `.lower()`ing of the data should be done in the model only where necessary.

This sounds right to me.  I don't think we should be manipulating user input without good reason, and that would best be done within the model on a case-by-case basis.

> We could do away with the to_lower arg for build_lookup_from_csv and read_csv_to_dataframe entirely, and have them always make the column names lowercase.

So long as we're committing to this convention of table indexes using lowercase strings, then yes, it would follow that we could remove the `to_lower` parameter from the function.

> If we're standardizing all column names to lowercase, this should also apply to the names of columns in CSVs that the model outputs.

We should take a closer look at this!  It is a related problem, but it affects different audiences.

Standardizing internal table indexing is a strictly _internal_ convention, so members of the software team are the only ones likely to be affected by the change.  Changing fieldnames in outputs, however, is an _external_-facing change, and so we need to also consider who this change will affect and to what degree it will affect them.

Because of these considerations, I suggest that we defer this for the time being.  Could you create an issue for it so we can take a look at it later on?

Reply:
Thanks! Linked issue here: https://github.com/natcap/invest/issues/262

Reply:
> > Making all the values lowercase is never necessary, it is only needed for a couple of columns in a couple of inputs.
> 
> You're right, I don't think we can safely lowercase _all_ string table values, and we've probably just gotten lucky with the ones that have been lowercased! Case will often matter (depending on the OS) for filepaths provided via a table, for example.

You're both on point. I just got tripped up by this in the Rec model on linux, where filepaths are being lowercased by `utils.build_csv...`.